File: src/main/java/me/lordsaad/modeoff/api/PlotAssigningManager.java
package me.lordsaad.modeoff.api;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.teamwizardry.librarianlib.features.kotlin.JsonMaker;
import me.lordsaad.modeoff.Modeoff;
import org.jetbrains.annotations.Nullable;

import java.io.*;
import java.util.ArrayList;
import java.util.UUID;

/**
 * Created by LordSaad.
 */
public class PlotAssigningManager {

	public static PlotAssigningManager INSTANCE = new PlotAssigningManager();

	private File plotFile;

	private PlotAssigningManager() {
		File directory = Modeoff.INSTANCE.getProxy().getDirectory();
		if (!directory.exists()) {
			Modeoff.INSTANCE.getLogger().info(directory.getName() + " directory not found. Creating directory...");
			if (!directory.mkdirs()) {
				Modeoff.INSTANCE.getLogger().fatal("SOMETHING WENT WRONG! Could not create config directory " + directory.getName());
				return;
			}
			Modeoff.INSTANCE.getLogger().info(directory.getName() + " directory has been created successfully!");
		}

		File plotFile = new File(directory, "registered_plots.json");
		try {
			if (!plotFile.exists()) {
				Modeoff.INSTANCE.getLogger().info(plotFile.getName() + " file not found. Creating file...");
				if (!plotFile.createNewFile()) {
					Modeoff.INSTANCE.getLogger().fatal("SOMETHING WENT WRONG! Could not create config file " + plotFile.getName());
					return;
				}

				JsonObject obj = new JsonObject();
				obj.add("plots", new JsonArray());

				FileWriter writer = new FileWriter(plotFile);
				writer.write(JsonMaker.serialize(obj));
				writer.flush();
				writer.close();
				Modeoff.INSTANCE.getLogger().info(plotFile.getName() + " file has been created successfully!");
			}
		} catch (IOException e) {
			e.printStackTrace();
		}

		this.plotFile = plotFile;
	}

	public boolean isUUIDRegistered(UUID uuid) {
		JsonElement json;
		try {
			json = new JsonParser().parse(new FileReader(plotFile));
			if (json.isJsonObject() && json.getAsJsonObject().has("plots") && json.getAsJsonObject().get("plots").isJsonArray()) {
				JsonArray array = json.getAsJsonObject().getAsJsonArray("plots");
				for (JsonElement element : array) {
					if (element.isJsonObject() && element.getAsJsonObject().has("uuid")) {
						UUID fetched = UUID.fromString(element.getAsJsonObject().getAsJsonPrimitive("uuid").getAsString());
						if (uuid.equals(fetched)) return true;
					}
				}
			}
		} catch (FileNotFoundException e) {
			e.printStackTrace();
		}

		return false;
	}

	public int getNextAvailableID() {
		JsonElement json;
		try {
			json = new JsonParser().parse(new FileReader(plotFile));
			if (json.isJsonObject() && json.getAsJsonObject().has("plots") && json.getAsJsonObject().get("plots").isJsonArray()) {
				JsonArray array = json.getAsJsonObject().getAsJsonArray("plots");
				if (array.size() == 0) return 0;
				return array.size();
			}
		} catch (FileNotFoundException e) {
			e.printStackTrace();
		}
		return -1;
	}

	public boolean saveUUIDToPlot(UUID uuid, int plotID) {
		if (isUUIDRegistered(uuid)) return false;

		JsonElement json;
		try {
			json = new JsonParser().parse(new FileReader(plotFile));
			if (json.isJsonObject() && json.getAsJsonObject().has("plots") && json.getAsJsonObject().get("plots").isJsonArray()) {
				JsonArray array = json.getAsJsonObject().getAsJsonArray("plots");

				ArrayList<JsonObject> plots = new ArrayList<>();
				for (JsonElement element : array)
					if (element.isJsonObject()) plots.add(element.getAsJsonObject());

				JsonObject newPlot = new JsonObject();
				newPlot.addProperty("uuid", uuid.toString());
				newPlot.addProperty("id", plotID);

				JsonArray newArray = new JsonArray();
				for (JsonElement element : plots) {
					newArray.add(element);
				}
				newArray.add(newPlot);

				JsonObject obj = new JsonObject();
				obj.add("plots", newArray);

				try {
					FileWriter writer = new FileWriter(plotFile);
					writer.write(JsonMaker.serialize(obj));
					writer.flush();
					writer.close();
					Modeoff.INSTANCE.getLogger().info("registered plot " + plotID + " to uuid '" + uuid.toString() + "' successfully!");

					return true;
				} catch (IOException e) {
					e.printStackTrace();
				}
			}
		} catch (FileNotFoundException e) {
			e.printStackTrace();
		}

		return false;
	}

	public int getPlotForUUID(UUID uuid) {
		JsonElement json;
		try {
			json = new JsonParser().parse(new FileReader(plotFile));
			if (json.isJsonObject() && json.getAsJsonObject().has("plots") && json.getAsJsonObject().get("plots").isJsonArray()) {
				JsonArray array = json.getAsJsonObject().getAsJsonArray("plots");
				for (JsonElement element : array) {
					if (element.isJsonObject() && element.getAsJsonObject().has("id") && element.getAsJsonObject().has("uuid")) {
						if (UUID.fromString(element.getAsJsonObject().getAsJsonPrimitive("uuid").getAsString()).equals(uuid))
							return element.getAsJsonObject().getAsJsonPrimitive("id").getAsInt();
					}
				}
			}
		} catch (FileNotFoundException e) {
			e.printStackTrace();
		}

		return -1;
	}

	@Nullable
	public UUID getUUIDForPlot(int plotId) {
		JsonElement json;
		try {
			json = new JsonParser().parse(new FileReader(plotFile));
			if (json.isJsonObject() && json.getAsJsonObject().has("plots") && json.getAsJsonObject().get("plots").isJsonArray()) {
				JsonArray array = json.getAsJsonObject().getAsJsonArray("plots");
				for (JsonElement element : array) {
					if (element.isJsonObject() && element.getAsJsonObject().has("id") && element.getAsJsonObject().has("uuid")) {
						if (element.getAsJsonObject().getAsJsonPrimitive("id").getAsInt() == plotId)
							return UUID.fromString(element.getAsJsonObject().getAsJsonPrimitive("uuid").getAsString());
					}
				}
			}
		} catch (FileNotFoundException e) {
			e.printStackTrace();
		}
		return null;
	}
}

File: src/main/kotlin/info/modoff/modeoff/api/PlotAssigningManager.kt
package info.modoff.modeoff.api

import com.google.gson.JsonArray
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import com.teamwizardry.librarianlib.features.kotlin.serialize
import info.modoff.modeoff.Modeoff

import java.io.*
import java.util.ArrayList
import java.util.UUID

/**
 * Created by LordSaad.
 */
object PlotAssigningManager {
    private val plotFile: File

    init {
        val directory = Modeoff.proxy.directory
        if (!directory.exists()) {
            Modeoff.logger.info(directory.name + " directory not found. Creating directory...")
            if (!directory.mkdirs()) {
                Modeoff.logger.fatal("SOMETHING WENT WRONG! Could not create config directory " + directory.name)
            }
            Modeoff.logger.info(directory.name + " directory has been created successfully!")
        }

        val plotFile = File(directory, "registered_plots.json")
        try {
            if (!plotFile.exists()) {
                Modeoff.logger.info(plotFile.name + " file not found. Creating file...")
                if (!plotFile.createNewFile()) {
                    Modeoff.logger.fatal("SOMETHING WENT WRONG! Could not create config file " + plotFile.name)
                }

                val obj = JsonObject()
                obj.add("plots", JsonArray())

                val writer = FileWriter(plotFile)
                writer.write(obj.serialize())
                writer.flush()
                writer.close()
                Modeoff.logger.info(plotFile.name + " file has been created successfully!")
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }

        this.plotFile = plotFile
    }

    fun isUUIDRegistered(uuid: UUID): Boolean {
        val json: JsonElement
        try {
            json = JsonParser().parse(FileReader(plotFile))
            if (json.isJsonObject && json.asJsonObject.has("plots") && json.asJsonObject.get("plots").isJsonArray) {
                val array = json.asJsonObject.getAsJsonArray("plots")
                array
                    .filter { it.isJsonObject && it.asJsonObject.has("uuid") }
                    .map { UUID.fromString(it.asJsonObject.getAsJsonPrimitive("uuid").asString) }
                    .filter { uuid == it }
                    .forEach { return true }
            }
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        }

        return false
    }

    val nextAvailableID: Int
        get() {
            val json: JsonElement
            try {
                json = JsonParser().parse(FileReader(plotFile))
                if (json.isJsonObject && json.asJsonObject.has("plots") && json.asJsonObject.get("plots").isJsonArray) {
                    val array = json.asJsonObject.getAsJsonArray("plots")
                    if (array.size() == 0) return 0
                    return array.size()
                }
            } catch (e: FileNotFoundException) {
                e.printStackTrace()
            }

            return -1
        }

    fun saveUUIDToPlot(uuid: UUID, plotID: Int): Boolean {
        if (isUUIDRegistered(uuid)) return false

        val json: JsonElement
        try {
            json = JsonParser().parse(FileReader(plotFile))
            if (json.isJsonObject && json.asJsonObject.has("plots") && json.asJsonObject.get("plots").isJsonArray) {
                val array = json.asJsonObject.getAsJsonArray("plots")

                val plots = ArrayList<JsonObject>()
                for (element in array)
                    if (element.isJsonObject) plots.add(element.asJsonObject)

                val newPlot = JsonObject()
                newPlot.addProperty("uuid", uuid.toString())
                newPlot.addProperty("id", plotID)

                val newArray = JsonArray()
                for (element in plots) {
                    newArray.add(element)
                }
                newArray.add(newPlot)

                val obj = JsonObject()
                obj.add("plots", newArray)

                try {
                    val writer = FileWriter(plotFile)
                    writer.write(obj.serialize())
                    writer.flush()
                    writer.close()
                    Modeoff.logger.info("registered plot " + plotID + " to uuid '" + uuid.toString() + "' successfully!")

                    return true
                } catch (e: IOException) {
                    e.printStackTrace()
                }

            }
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        }

        return false
    }

    fun getPlotForUUID(playerUuid: UUID): Int {
        try {
            val json = JsonParser().parse(FileReader(plotFile))
            (json as? JsonObject)?.let {
                (it["plots"] as? JsonArray)
                    ?.filterIsInstance<JsonObject>()
                    ?.forEach {
                        val id = it["id"]
                        val uuid = it["uuid"]
                        if (id != null && uuid != null) {
                            if (UUID.fromString(uuid.asString) == playerUuid) {
                                return id.asInt
                            }
                        }
                    }
            }
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        }

        return -1
    }

    fun getUUIDForPlot(plotId: Int): UUID? {
        try {
            val json = JsonParser().parse(FileReader(plotFile))
            (json as? JsonObject)?.let {
                (it["plots"] as? JsonArray)
                    ?.filterIsInstance<JsonObject>()
                    ?.forEach {
                        val id = it["id"]
                        val uuid = it["uuid"]
                        if (id != null && uuid != null) {
                            if (id.asInt == plotId) {
                                return UUID.fromString(uuid.asString)
                            }
                        }
                    }
            }
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        }

        return null
    }
}
